### ISSUE TITLE
Unsupported readline functions 'menu-complete', 'menu-complete-backward', 'vi-undo' and 'backward-kill-word'

### ISSUE THREAD

**Issue by excited-bore** [akinomyoga/ble.sh#534]
**ble version**: 0.4.0-devel4+31f264ad
**Bash version**: GNU bash, version 5.2.37(1)-release (x86_64-pc-linux-gnu)

Upon loading ble.sh for the first time, I'm greeted with these errors:
```
ble.sh (bind): unsupported readline function 'menu-complete' for keymap 'vi_nmap'.
ble.sh (bind): unsupported readline function 'menu-complete-backward' for keymap 'vi_nmap'.
ble.sh (bind): unsupported readline function 'vi-undo'.
ble.sh (bind): unsupported readline function 'vi-undo'.
ble.sh (bind): unsupported readline function 'backward-kill-word' for keymap 'vi_nmap'.
```
Installed from the AUR, added both suggested lines at the top and bottom of my ~/.bashrc

**Comment by akinomyoga** [comment]
They are unsupported.

Do you have those bindings in your `~/.inputrc` or in `~/.bashrc`? As explained in the message, they are unsupported in those keymaps. If you really need them, I can implement them, but it will take time because I'm busy.

- The support for `menu-complete` and `menu-complete-backward` in the vi-normal mode might be added.
- `vi-undo` is a relatively new feature of Bash introduced in the latest version of Bash 5.2, and it's not yet been supported. Actually, in ble.sh, `undo` in the vi-normal keymap is a synonym for `vi-undo`. You can simply use `undo`.
- `backward-kill-word` in the vi-normal mode is subtle.
  - It is non-trivial how this binding should behave in the vi-style editing using vi-motion and operators. There are many things to consider. We can probably make it work with the arguments (i.e., prefixed numbers) and registers. The registers `[` and `]` would also be well-defined. How should the subsequent `.` after "backward-kill-word" behave?
  - It might be considered as a combination such as e.g. `db`, but then why don't you just bind `"db"`?

**Comment by excited-bore** [comment]
These binding where part of my bashconfig. I'm sorry to hear you're busy and I can definitely understand that reimplementing these readline functions is nothing less then quite a laborious task. 

If you find the time for it, these functions I _personally_ look forward to the most:
`history-and-alias-expand-line` : To expand an alias on the line
`vi-undo` (or a basic way that `vi-undo` gets translated into `undo`,

There are some other ones that I use that seem usefull to add in the longterm - `forward/backward-word`, `vi-forward/backward-(big)word`, `(shell)-backward/forward-kill-word`, `quoted-insert` and `re-read-init-file`, but I understand that this is probably more of and undertaking then led on. 
And it's understandably not always as necessary as with `menu-complete(-backward)` or `emacs/vi-editing-mode` since the alternatives ble.sh supplies are some essentially are sometimes essentially just straight up upgrades (at least with menucompletion stuff) which make some of these rl-function completely obsolete. 

Most of the readline stuff are niceties though that can take time being imported. Otoh `vi-undo` not being supported because there's only `undo` feels like an very preventable compatibility issue. 
It's really annoying having to work with multiple shellconfigs, and the specific variants of those configs as well - in case I can't install ble.sh on something for whatever reason, like on the kind of server where nothing can break - so I'd mostly be nice if you could make sure there's a `vi-undo` variant that does the same as `undo` just for compatibility's sake.

Thank you for this project and taking the time to read through all this, even though the incompatiability with some of the readline functions is a little annoying, it does really look interesting and Im definitely going to be using this on and off.

**Comment by akinomyoga** [comment]
What version of `ble.sh` do you use? You reported "ble version: 0.4.0-devel4+31f264ad" at the beginning, but I now start to doubt it. How did you check the version?

What is the result of running the command `ble/widget/display-shell-version`?

**Comment by excited-bore** [comment]
Ah, got me. AUR package is outdated.
It's on version 0.3.4+566651e

**Comment by akinomyoga** [comment]
So what readline functions are missing with `blesh-git` (i.e., the latest version)?

`menu-complete{,-backward}` in `vi_nmap` is not supported. Those readline functions should basically be the same as in `vi_imap`, so I guess I just need to slightly adjust the behavior of the one for `vi_imap`.

For rl-function `vi-undo`, I tried to understand what would be the difference from rl-function `undo`, but it seems just a synonym even in the Bash source code:

```console
$ git grep '\bundo"'
lib/readline/funmap.c:  { "undo", rl_undo_command },
lib/readline/funmap.c:  { "vi-undo", rl_vi_undo },
[murase@chatoyancy 1 bash-dev]$ g grep -C 3 '^rl_vi_undo'
lib/readline/vi_mode.c-
lib/readline/vi_mode.c-/* A placeholder for further expansion. */
lib/readline/vi_mode.c-int
lib/readline/vi_mode.c:rl_vi_undo (int count, int key)
lib/readline/vi_mode.c-{
lib/readline/vi_mode.c-  return (rl_undo_command (count, key));
lib/readline/vi_mode.c-}
```

The readline bindable function `vi-undo` is implemented as a C function `rl_vi_undo`, but this function simply calls `rl_undo_command`, which is actually the readline bindable function `undo`.

`backward-kill-word` needs time to consider the proper behavior in `vi_nmap`.

**Comment by excited-bore** [comment]
The four ones from the initial post: `'menu-complete', 'menu-complete-backward', 'vi-undo', 'backward-kill-word'`

**Comment by akinomyoga** [comment]
OK, thanks for the clarification.

**Comment by akinomyoga** [comment]
I added support for these readline bindable functions in commit d7ec488a7191c490a22fe46dd1129306a690ee98. Could you update `ble.sh` by running `ble-update` and see the behavior?

**Comment by excited-bore** [comment]
```
ble.sh (bind): warning: empty keyseq
ble.sh (bind): warning: empty keyseq
ble.sh (bind): warning: empty keyseq
ble.sh (bind): warning: empty keyseq
ble.sh (bind): warning: empty keyseq
ble.sh (bind): warning: empty keyseq
```
This is what I get.

**Comment by excited-bore** [comment]
```
ble.sh (bind): warning: empty keyseq
ble.sh (bind): warning: empty keyseq
ble.sh (bind): warning: empty keyseq
ble.sh (bind): warning: empty keyseq
ble.sh (bind): warning: empty keyseq
ble.sh (bind): warning: empty keyseq
```
This is what I get.
It seems to be related to a file i source in my ~/.bashrc. Or to say, a file I source *by sourcing another file*, which seems to be the direct problem for this

Either way when I source the file directly to my ~/.bashrc, the error together with all other error

**Comment by akinomyoga** [comment]
> ```
> ble.sh (bind): warning: empty keyseq
> ```

Hmm, this should only happen when one specifies `bind '"": something'` (where the key sequence enclosed within `""` is empty). `ble.sh` issues an error message for it on purpose because there is no use case for `bind '"": something'` as far as I know.

Do you have settings like `bind '"": something'`? Or do you think other forms of `bind '...'` trigger this error?

**Comment by akinomyoga** [comment]
> Either way when I source the file directly to my ~/.bashrc, the error together with all other errors

So the issue happens with the way how you source the settings? Hmm, this should not happen. If possible, I'd like to fix it. Could you provide the actual contents of the settings, where the issue happened?

**Comment by akinomyoga** [comment]
I extracted the key sequences used in [`dotfiles/keybinds/.keybinds.d/keybinds.bash` at 4d3a679 · `excited-bore/dotfiles`](https://github.com/excited-bore/dotfiles/blob/4d3a679daad3cb1741305824b8635ec2b066c93c/keybinds/.keybinds.d/keybinds.bash):

```bash
bind '"'\''": bell'
bind '"\201": bell'
bind '"\202": bell'
bind '"\205": bell'
bind '"\206": bell'
bind '"\207": bell'
bind '"\208": bell'
bind '"\209": bell'
bind '"\C-b": bell'
bind '"\C-d": bell'
bind '"\C-e": bell'
bind '"\C-e\C-e": bell'
bind '"\C-f": bell'
bind '"\C-g": bell'
bind '"\C-h": bell'
bind '"\C-l": bell'
bind '"\C-o": bell'
bind '"\C-q": bell'
bind '"\C-s": bell'
bind '"\C-v": bell'
bind '"\C-w": bell'
bind '"\C-x'\''": bell'
bind '"\C-x\C-b": bell'
bind '"\C-x\C-j": bell'
bind '"\C-z": bell'
bind '"\e266": bell'
bind '"\e277": bell'
bind '"\e288": bell'
bind '"\eOQ": bell'
bind '"\eOR": bell'
bind '"\eOS": bell'
bind '"\e[15~": bell'
bind '"\e[17~": bell'
bind '"\e[18~": bell'
bind '"\e[19~": bell'
bind '"\e[1;2A": bell'
bind '"\e[1;2B": bell'
bind '"\e[1;2C": bell'
bind '"\e[1;2D": bell'
bind '"\e[1;3A": bell'
bind '"\e[1;3B": bell'
bind '"\e[1;3C": bell'
bind '"\e[1;3D": bell'
bind '"\e[1;5A": bell'
bind '"\e[1;5B": bell'
bind '"\e[1;5C": bell'
bind '"\e[1;5D": bell'
bind '"\e[9;5u": bell'
bind '"\e[9;5u": bell'
bind '"\e[A": bell'
bind '"\e[B": bell'
bind '"\e[Z": bell'
bind '"v": bell'
bind 'Tab: bell'
```

Then, I tried to source this file, but the problem of "empty keyseq" did not seem to reproduce. I also tried to source the file via another sourced file, but the problem didn't reproduce either. Do you have any advice to reproduce the behavior?

**Comment by excited-bore** [comment]
I have file `~/.keybinds` that sources every bash script related to keybindings  

This seems to be the root cause of the problem:
```bash
#!/bin/bash

if [ -d ~/.keybinds.d/ ] && [ "$(ls -A ~/.keybinds.d/)" ]; then
  for keybind in ~/.keybinds.d/*.bash; do
      source "$keybind" 
  done
fi

```

**Comment by akinomyoga** [comment]
Thanks. I have questions.

**Q1**: I see only `keybinds.bash` in your repository, but do you have other additional files in your `.keybinds.d` locally?

**Q2**: Does the problem reproduce if you temporarily replace the content of your `keybinds.bash` with the code block in https://github.com/akinomyoga/ble.sh/issues/534#issuecomment-2561534494?

**Comment by excited-bore** [comment]
You were right in https://github.com/akinomyoga/ble.sh/issues/534#issuecomment-2561502638 and it was good that you made me check whether I had multiple files in my `~/.keybinds.d/` folder because I did have 2 symlinks to the same problematic file from [fzf](https://github.com/junegunn/fzf); `~/.fzf/shell/key-bindings.bash` 
I tend to make scripts that use `sed -i 's|..|..|g'`  and I probably composed a poor sed while testing because yes, there were just a few lines that had put `bind '"": something'` in that file... 

```bash
    # CTRL-F - Search with previews and other handy additions
    bind -m emacs-standard -x '"": fzf_rifle'
    bind -m vi-command -x '"": fzf_rifle'
    bind -m vi-insert -x '"":  fzf_rifle'
```

Apparently default bash's bind just doesn't even bother reporting errors like these huh.. Now I know I guess

So TL;DR the `ble.sh (bind): warning: empty keyseq`'s are not ble.sh bugs, they're bugs related to my personal config (sorry)
Thanks for being so thorough on stuff like this though, it's really commendable. :+1: 
(And at least I know now that if I ever see that come up again I should double check if my 'install_fzf.sh' script works properly?)

**Comment by akinomyoga** [comment]
> So TL;DR the `ble.sh (bind): warning: empty keyseq`'s are not ble.sh bugs, they're bugs related to my personal config (Sorry!

Never mind. I decided to include the filename and the line number in the error message in commit 89c69077305b5c9da421720879f4fdb40b437907, so that the user knows where to look at. This should be useful when debugging their keybinding settings.
